### DISCUSSION TITLE
SDL limit FPS in Fullscreen mode

### DISCUSSION THREAD

**Discussion by MrScautHD** [aquagoose/Pie#39] [Help and Questions]
Hello, by setting the game into fullscreen, the fps get limit to 60 FPS.

Info:
With SDL D3D11 and OpenGL get limited!

**Comment by aquagoose** [comment]
Make sure vertical sync is disabled (set swap interval to 0) and is not forced on by your driver. Apart from that, this will most likely be driver & implementation specific for which I can do nothing about.

**Comment by MrScautHD** [comment]
![grafik](https://github.com/piegfx/Pie/assets/65916181/60efabd9-9989-4878-9f08-9c9732c6fa68)

By disabling it does work but it should possible to override this variable and set it to false.

**Comment by aquagoose** [comment]
I'm assuming your global setting is to have vertical sync enabled? It is not possible to override this value, as a global value is forced by the driver, regardless of the application's settings. You should select "use application setting".

**Comment by MrScautHD** [comment]
"use application setting" does not exist, and what use other games to fix this issue?

**Comment by aquagoose** [comment]
There should be a setting similar to it. I haven't used the nvidia control panel in a long time so I don't remember the exact name. This sounds to me like a driver setting.

**Comment by MrScautHD** [comment]
![grafik](https://github.com/piegfx/Pie/assets/65916181/1a234df4-63dd-43e5-ba7a-fba6fc310f78)

Oh, sry it is wrong translaited if i set it to it and it does not work.

**Comment by aquagoose** [comment]
I will have to test on a windows machine. However I don't think there is any solution to this problem.

**Comment by MrScautHD** [comment]
> 
And SDL does not have any method to disable VSync or any other method to disable it...

**Comment by MrScautHD** [comment]
i found something: https://www.gamedev.net/forums/topic/665862-how-to-disable-vsync-in-sdl-2/

**Comment by aquagoose** [comment]
SDL does itself not handle vertical sync at all. SDL just interfaces with the underlying window system.

**Comment by MrScautHD** [comment]
> SDL does itself not handle vertical sync at all. SDL just interfaces with the underlying window system.

But what is that? https://wiki.libsdl.org/SDL2/SDL_GL_SetSwapInterval

**Comment by aquagoose** [comment]
`SDL_GL_SetSwapInterval` will interface with the various underlying OpenGL systems. For windows, it will call `wglSwapIntervalEXT`, for X11 it will call `glxSetSwapInterval` etc.

**Comment by MrScautHD** [comment]
> `SDL_GL_SetSwapInterval` will interface with the various underlying OpenGL systems. For windows, it will call `wglSwapIntervalEXT`, for X11 it will call `glxSetSwapInterval` etc.

well but GL has the same problem

**Comment by aquagoose** [comment]
As I've said, I have no control over what SDL does when setting the swap interval. However since all the function does is call the window manager functions, I have even less control as those functions are directly handled by the window manager and display driver.

**Comment by MrScautHD** [comment]
> 

Did you googled it already if any user has the same issue?

**Comment by aquagoose** [comment]
Sorry?

**Comment by MrScautHD** [comment]
> Sorry?

No i mean maybe there solutions, i tried already to look. (but i just know Stackoverflow)

**Comment by MrScautHD** [comment]
That is what i found:

DirectX11
```cs
// Create a Device object for DirectX
Device device = new Device(DriverType.Hardware, DeviceCreationFlags.None);

// Get the DXGI device associated with the DirectX device
using (var dxgiDevice = device.QueryInterface<SharpDX.DXGI.Device>())
{
    // Get the DXGI output associated with the primary display
    using (var output = dxgiDevice.Adapter.GetOutput(0))
    {
        // Get the DXGI factory associated with the DXGI output
        using (var dxgiFactory = output.GetParent<SharpDX.DXGI.Factory>())
        {
            // Disable VSync for the DXGI factory
            dxgiFactory.MakeWindowAssociation(System.Diagnostics.Process.GetCurrentProcess().MainWindowHandle, WindowAssociationFlags.None);
        }
    }
}
```

OpenGL:
```cs
// Disable VSync for OpenGL
private void DisableVSync()
{
    // Get the current OpenGL context
    var context = new OpenTK.Platform.Utilities.GraphicsContextHandle(OpenTK.Graphics.GraphicsContext.CurrentContext.Context.Handle);
    
    // Disable VSync
    if (context != OpenTK.Platform.Utilities.GraphicsContextHandle.Zero)
    {
        if (context.IsCurrent)
        {
            // Set the swap interval to 0 (disable VSync)
            GL.GetDelegateForExtension<GL.Arb>().SwapInterval(0);
        }
        else
        {
            MessageBox.Show("Failed to disable VSync. The OpenGL context is not current.");
        }
    }
    else
    {
        MessageBox.Show("Failed to disable VSync. No valid OpenGL context found.");
    }
}
```

**Comment by aquagoose** [comment]
Assuming that the swap interval is actually 0, for OpenGL there is no solution at all. It is driver specific. For Direct3D, you can force it to disable vertical sync by setting a swapchain parameter. However this is beyond the scope of what Pie will implement, as it is not in any way compatible with OpenGL, and Pie does not implement anything that is not compatible with a backend.

However this swapchain parameter also forces exclusive fullscreen which is mostly recommended that you avoid.

**Comment by MrScautHD** [comment]
oh :/

I see you working on Vulkan, would that has the same issue?

**Comment by aquagoose** [comment]
A driver or window manager may prevent a vulkan swapchain from using anything but a FIFO presentation mode (vertical sync). However Pie itself will, assuming vertical sync is disabled, attempt to request a mailbox (triple buffering), relaxed fifo (adaptive sync), or immediate (no vertical sync at all) presentation mode depending on the swap interval chosen. However it is down to the driver to actually provide this, and if it does not allow the user to use a certain option, Pie will pick the next best available option, ultimately resorting to vertical sync if no other option is available.

**Comment by MrScautHD** [comment]
so it is not fixable for Pie?

**Comment by aquagoose** [comment]
No, as this sounds like a driver related issue, especially since you are getting the same issue with OpenGL (this is however a known "feature" of Direct3D, with the only solution being to set a swapchain parameter, which is beyond the scope of Pie).

I would suggest however making sure your swap interval is indeed 0 as if it is >1 then vertical sync will be enabled.

**Comment by MrScautHD** [comment]
how can i check this?

**Comment by aquagoose** [comment]
Check the value that is being passed into `GraphicsDevice.Present`.

If you want to go a step further, you can also check the value that is being passed into `IDXGISwapchain.Present` for the Direct3D device, and `SDL_GL_SetSwapInterval` for the OpenGL device.

**Comment by aquagoose** [comment]
I will check on Windows now to see if there is anything I have missed however I have tried to fix this issue with Direct3D and have been unable to. However I did last check with the old windowing system so there is a chance things may have changed since I implemented SDL, and there is a chance that things that did not work before will now start working.

**Comment by aquagoose** [comment]
I've added a couple of swapchain parameters in f00802c that should allow for vertical sync to be disabled in borderless fullscreen windows, if the driver allows it. I've done extensive tests and OpenGL and Direct3D behave the same for both windowed, borderless, and exclusive fullscreen windows.

**Comment by MrScautHD** [comment]
Nice! but i can just try it if Pie get updated in Easel.

**Comment by aquagoose** [comment]
has been done.

**Comment by MrScautHD** [comment]
nice thx!

**Comment by MrScautHD** [comment]
> I've added a couple of swapchain parameters in [f00802c](https://github.com/piegfx/Pie/commit/f00802c3aa050201029e86e4a6336931665ecf5f) that should allow for vertical sync to be disabled in borderless fullscreen windows, if the driver allows it. I've done extensive tests and OpenGL and Direct3D behave the same for both windowed, borderless, and exclusive fullscreen windows.

need i to set something or should it work automaticlly?

**Comment by aquagoose** [comment]
It will work automatically.

**Comment by MrScautHD** [comment]
well then it does not work :/

**Comment by aquagoose** [comment]
Have you checked to make sure that the swap interval is actually 0?

Try cloning [Pie.Tests](https://github.com/piegfx/Pie/tree/master/tests/Pie.Tests) and running it to see if the issue persists. It prints the current FPS to the console every second. You can experiment with setting the swap interval in `Present`, and also try changing the fullscreen mode when F11 is pressed.

**Comment by MrScautHD** [comment]
![grafik](https://github.com/piegfx/Pie/assets/65916181/9ba74dfe-68b2-47eb-ae1d-8e42629679ad)

It is 0, yes

**Reply by aquagoose** [reply]
Please try the `Pie.Tests` that I suggested in the previous comment to make sure this is not an easel issue.

**Reply by MrScautHD** [reply]
sure let me fork it.

**Reply by MrScautHD** [reply]
Does it just work for DirectX11?

**Reply by MrScautHD** [reply]
Because i tried DirectX11 and it crash and by OpenGl it get limited even in Window mode to 60 FPS
![grafik](https://github.com/piegfx/Pie/assets/65916181/e630f931-ebe2-43cc-9e8e-61c75aecb30a)

**Reply by aquagoose** [reply]
The Direct3D crash is likely because it creates a debug graphics device. Try disabling `debug` in the `GraphicsDeviceOptions`. As for OpenGL, I have tested it yet again and it does indeed run at an unlocked framerate.

**Reply by MrScautHD** [reply]
Well i forgot to set it to `GraphicsDevice.Present(0);`

OpenGl works now fine

DirectX11 still has the problem

**Comment by MrScautHD** [comment]
Idk why but OpenGl works without any bugs!
```cs
Window = Fine
ExclusiveFullscreen = Fine
BorderlessFullscreen = Fine
```

But DirectX11 has still this wierd issue
```cs
Window = Fine
ExclusiveFullscreen = broken
BorderlessFullscreen = broken
```

**Reply by MrScautHD** [reply]
So any idea to fix it on DirectX11 to ;)

**Reply by aquagoose** [reply]
I have been unable to recreate your issue on every device I have tried. I can't fix an issue that from what I can see works correctly I'm afraid. I suspect this is an issue on your end.

**Reply by MrScautHD** [reply]
Does it work for you on DirectX11?

**Reply by aquagoose** [reply]
Yes.

**Reply by MrScautHD** [reply]
What Device do you have?

I have a Gtx 1080

**Reply by aquagoose** [reply]
I have tested with both Intel and Nvidia.

**Reply by MrScautHD** [reply]
:/

**Reply by MrScautHD** [reply]
And you has no other idea to fix this?

How did monogame fixed this issue, i checked it there and it works fine...

**Reply by aquagoose** [reply]
monogame by default uses exclusive fullscreen, and it also directly interfaces with win32. I suspect if you disable `HardwareModeSwitch` you'll get the same issue.

**Reply by MrScautHD** [reply]
Mh, so any idea then?

**Reply by MrScautHD** [reply]
Or is it just no fixable with DirectX then... :/
